pkg/network/setup/network_suite_test.go

```diff
@@ -16,12 +16,12 @@ func TestNetwork(t *testing.T) {
 func newVMI(namespace, name string) *v1.VirtualMachineInstance {
 	vmi := api2.NewMinimalVMIWithNS(namespace, name)
 	vmi.Spec.Networks = []v1.Network{*v1.DefaultPodNetwork()}
+	vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{*v1.DefaultBridgeNetworkInterface()}
 	return vmi
 }
 
 func newVMIBridgeInterface(namespace string, name string) *v1.VirtualMachineInstance {
 	vmi := newVMI(namespace, name)
-	vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{*v1.DefaultBridgeNetworkInterface()}
```

Comment:
this line is fishy: if you remove this line from the `newVMIBridgeInterface` function, what is creating the bridge interface on the VMI ? 
Is it created by the function that sets the defaults - which is invoked below ? 

I'm also having trouble matching this change with the content of the commit msg.

Reply:
> this line is fishy: if you remove this line from the `newVMIBridgeInterface` function, what is creating the bridge interface on the VMI ? Is it created by the function that sets the defaults - which is invoked below ?

It is created by the function that is called before (`newVMI`) which I edited to also include the interface.

> I'm also having trouble matching this change with the content of the commit msg.

The change in `newVMI` was a direct sub-sequence of the 1:1 relation that is expected in the production and mentioned in the commit message. I did it because without it, the test does not pass.

Reply:
> > this line is fishy: if you remove this line from the `newVMIBridgeInterface` function, what is creating the bridge interface on the VMI ? Is it created by the function that sets the defaults - which is invoked below ?
> 
> It is created by the function that is called before (`newVMI`) which I edited to also include the interface.
> 
> > I'm also having trouble matching this change with the content of the commit msg.
> 
> The change in `newVMI` was a direct sub-sequence of the 1:1 relation that is expected in the production and mentioned in the commit message. I did it because without it, the test does not pass.

What doesn't make sense to me is that we now have 2 functions: 
  - `newVMI`: creates a vmi with a bridge interface.
  - `newVMIBridgeInterface`: creates a vmi with a bridge interface.

I find this hard to reason about; in my own personal mental model, this was correct before, since we had:
  - `newVMI`: creates a vmi without interfaces.
  - `newVMIBridgeInterface`: creates a vmi with a bridge interface.

What was wrong before ?... AFAIU, having no networks but also no interfaces in the VMI spec is correct... Which is what the `newVMI` function created.

Reply:
I have hard time to reason for the existing tests, but I cannot fix it all (although I want to, really).
I would have removed the function if `v1.SetObjectDefaults_VirtualMachineInstance(vmi)` was not there.
I basically did the minimum reasonable to allow everything to work and still make sense.

> What was wrong before ?... AFAIU, having no networks but also no interfaces in the VMI spec is correct... Which is what the newVMI function created.

It created a default network without an interface, that is the problem that had to be fixed.

Reply:
> I have hard time to reason for the existing tests, but I cannot fix it all (although I want to, really). I would have removed the function if `v1.SetObjectDefaults_VirtualMachineInstance(vmi)` was not there. I basically did the minimum reasonable to allow everything to work and still make sense.
> 
> > What was wrong before ?... AFAIU, having no networks but also no interfaces in the VMI spec is correct... Which is what the newVMI function created.
> 
> It created a default network without an interface, that is the problem that had to be fixed.

Perfect, now I understand. Tks.

IMO, the fix should instead make `newVMI` have **no** networks & **no** interfaces, and `newVMIBridgeInterface` should have **one** network & **one** corresponding interface, of type bridge.

It's very hard to chew the fact that 2 funcs - `newThing` and `newThingWithOtherSmallerThingInside` - both return a "thing with other smaller thing inside".

Reply:
What you are asking makes sense if you focus on the tests, but if you consider that changing tests is bad when you need to change the production code (it is suppose to protect you after-all), it makes it less appealing.
Changing to the logic above will cause a chain of other changes in the tests, which I hoped to avoid.

As you insist, I am forced to do these changes *before* this first commit, so it will have its own context and reasoning.

Reply:
I probably misunderstood your comment, but what changes would this require to the production code ? 

AFAIU, the `newVMI` and `newVMIBridgeInterface` functions are just some un-exported test helpers defined in the suite's test file: https://github.com/kubevirt/kubevirt/blob/83508a86f29e94dad48c698870f93f465f78cf4f/pkg/network/setup/network_suite_test.go#L16

Reply:
Changes in tests beyond what the current commit did.

Reply:
Done